File: 2262-solving-questions-with-brainpower/solving-questions-with-brainpower.cpp
class Solution {
public:
    long long mostPoints(vector<vector<int>>& questions) {
        /*
            dp[i] = max points to solve from index `i` to end
            dp[i] = max( dp[i + 1] `leave`,  points[i] + dp[i + skip[i]] `take` )

            // basecase
                dp[n] = points[n]
        */

        int const n = static_cast<int>(questions.size());
        vector dp(n, 0LL);

        dp[n - 1] = questions[n - 1][0];

        for (int i = n - 2; ~i ; i--) {
            int points = questions[i][0];
            int skip   = questions[i][1] + 1;
            
            int64_t leave = dp[i + 1];
            int64_t take = points;
            if ( i + skip < n )
                take = take + dp[i + skip];

            dp[i] = max(take, leave);
        }

        return dp[0];
    }
};

File: 2262-solving-questions-with-brainpower/solving-questions-with-brainpower.cs
/*
 q[i][0]

 dp[i] -> max points you can get if you start from index `i` to end

 dp[i] -> max( dp[i + 1], q[i][0] + dp[q[i][1] + i + 1] )

 dp[n] -> q[n][0]
 */

public class Solution
{
    public long MostPoints(int[][] q)
    {
        int n = q.Length;

        var dp = new long[n];

        // basecase
        dp[n - 1] = q[n - 1][0];

        for (int i = n - 2; i > -1 ; i --)
        {

            long leave = dp[i + 1];

            long take = q[i][0];

            int nxt = q[i][1] + i + 1;
            if( nxt < n )
            {
                take = take + dp[nxt];
            }

            dp[i] = long.Max(take, leave);
        }

        return dp[0];
    }
}